Mathlib/Analysis/SpecialFunctions/Complex/Circle.lean

```diff
@@ -235,3 +236,47 @@ open AddCircle
 lemma isLocalHomeomorph_expMapCircle : IsLocalHomeomorph expMapCircle := by
   have : Fact (0 < 2 * π) := ⟨by positivity⟩
   exact homeomorphCircle'.isLocalHomeomorph.comp (isLocalHomeomorph_coe (2 * π))
+
+namespace ZMod
+
+/-!
+### Additive characters valued in the complex circle
+-/
+
+open scoped Real
+
+variable {N : ℕ} [NeZero N]
+
+/-- The additive character from `ZMod N` to the unit circle in `ℂ`, sending `j mod N` to
+`exp (2 * π * I * j / N)`. -/
+noncomputable def toCircle : AddChar (ZMod N) circle where
+  toFun := fun j ↦ (toAddCircle j).toCircle
+  map_add_eq_mul' a b := by simp_rw [map_add, AddCircle.toCircle_add]
+  map_zero_eq_one' := by simp_rw [map_zero, AddCircle.toCircle, ← QuotientAddGroup.mk_zero,
+    Function.Periodic.lift_coe, mul_zero, expMapCircle_zero]
+
+lemma toCircle_intCast (j : ℤ) :
+    toCircle (j : ZMod N) = Complex.exp (2 * π * Complex.I * j / N) := by
+  rw [toCircle, AddChar.coe_mk, AddCircle.toCircle, toAddCircle_intCast,
+    Function.Periodic.lift_coe, expMapCircle_apply]
+  push_cast
+  ring_nf
+
+lemma toCircle_natCast (j : ℕ) :
+    toCircle (j : ZMod N) = Complex.exp (2 * π * Complex.I * j / N) := by
+  rw [← Int.cast_natCast, toCircle_intCast, Int.cast_natCast]
```

Comment:
```suggestion
  simpa using toCircle_intCast j
```
is a more robust proof

Reply:
The "more robust proof" does not quite work because it cannot infer `N` in the argument. I went for `simpa using toCircle_intCast (N := N) j`.